### PR TITLE
Changes for unique emails

### PR DESCRIPTION
**Pull request by dbkr**
Tighten unique constraint on user_threepids
Add endpoint proxying ID server requestEmailToken, erroring if email in use

Schema version has NOT been bumped.

### CODE REVIEW

**Comment by erikjohnston** [comment]
> Schema version has NOT been bumped.

Why?

**Comment by dbkr** [comment]
Just wasn't sure whether that was desired or not, so just noting that I hadn't done it.

**Comment by erikjohnston** [comment]
> Just wasn't sure whether that was desired or not, so just noting that I hadn't done it.

I think its best just to bump it, otherwise its likely to get forgotten. That and nothing will test it until the version gets bumped.

**Comment by dbkr** [comment]
Bumped

**Comment by erikjohnston** [comment]
LGTM

**Comment by NegativeMjark** [comment]
LGTM

File: synapse/rest/client/v2_alpha/register.py [hunk 1]
```diff
@@ -41,7 +41,7 @@
 
 
 class RegisterRestServlet(RestServlet):
-    PATTERN = client_v2_pattern("/register")
+    PATTERN = client_v2_pattern("/register*")
```
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/register.py]: Is this intended? I thought these were regexps. "registerrrrrrrr"

File: synapse/rest/client/v2_alpha/register.py [hunk 2]
```diff
@@ -209,6 +214,29 @@ def _create_registration_details(self, user_id, token):
...
+        body = parse_json_dict_from_request(request)
+
+        required = ['id_server', 'client_secret', 'email', 'send_attempt']
+        absent = []
+        for k in required:
+            if k not in body:
+                absent.append(k)
+
+        existingUid = yield self.hs.get_datastore().get_user_id_by_threepid(
+            'email', body['email']
```
Comment (NegativeMjark) [inline synapse/rest/client/v2_alpha/register.py]: Should be be accessing keys in the body before checking the ``absent`` list?

File: synapse/rest/client/v2_alpha/register.py (hunk 1 above)
Comment (dbkr) [inline synapse/rest/client/v2_alpha/register.py]: Thank you!

**Pull request merged**